pkg/proxy/healthcheck/proxier_health.go

```diff
@@ -34,110 +34,71 @@ const (
 
-// NewProxierHealthServer returns a proxier health http server.
-func NewProxierHealthServer(addr string, healthTimeout time.Duration) ProxierHealthUpdater {
-	return newProxierHealthServer(stdNetListener{}, stdHTTPServerFactory{}, clock.RealClock{}, addr, healthTimeout)
+// NewProxyHealthzServer returns a proxier health http server.
+func NewProxyHealthzServer(addr string, healthTimeout time.Duration) *ProxyHealthzServer {
+	return newProxyHealthzServer(stdNetListener{}, stdHTTPServerFactory{}, clock.RealClock{}, addr, healthTimeout)
 }
 
-func newProxierHealthServer(listener listener, httpServerFactory httpServerFactory, c clock.Clock, addr string, healthTimeout time.Duration) *proxierHealthServer {
-	hs := &proxierHealthServer{
-		listener:      listener,
-		httpFactory:   httpServerFactory,
-		clock:         c,
-		addr:          addr,
-		healthTimeout: healthTimeout,
+func newProxyHealthzServer(listener listener, httpServerFactory httpServerFactory, c clock.Clock, addr string, healthTimeout time.Duration) *ProxyHealthzServer {
+	hs := &ProxyHealthzServer{
+		listener:              listener,
+		httpFactory:           httpServerFactory,
+		clock:                 c,
+		addr:                  addr,
+		healthTimeout:         healthTimeout,
+		proxierHealthManagers: make([]*ProxierHealthManager, 0),
 	}
+
 	// The node is eligible (and thus the proxy healthy) while it's starting up
 	// and until we've processed the first node event that indicates the
 	// contrary.
 	hs.nodeEligible.Store(true)
 	return hs
 }
 
-// Updated indicates that kube-proxy has successfully updated its backend, so it should
-// be considered healthy now.
-func (hs *proxierHealthServer) Updated() {
-	hs.oldestPendingQueued.Store(zeroTime)
-	hs.lastUpdated.Store(hs.clock.Now())
-}
-
-// QueuedUpdate indicates that the proxy has received changes from the apiserver but
-// has not yet pushed them to its backend. If the proxy does not call Updated within the
-// healthTimeout time then it will be considered unhealthy.
-func (hs *proxierHealthServer) QueuedUpdate() {
-	// Set oldestPendingQueued only if it's currently zero
-	hs.oldestPendingQueued.CompareAndSwap(zeroTime, hs.clock.Now())
+// NewProxierHealthManager instantiates a new ProxierHealthManager and adds it to proxierHealthManagers.
+func (hs *ProxyHealthzServer) NewProxierHealthManager() *ProxierHealthManager {
+	hm := newProxierHealthManager(hs.clock, hs.healthTimeout)
+	hs.proxierHealthManagers = append(hs.proxierHealthManagers, hm)
+	return hm
 }
 
-// IsHealthy returns only the proxier's health state, following the same
-// definition the HTTP server defines, but ignoring the state of the Node.
-func (hs *proxierHealthServer) IsHealthy() bool {
-	isHealthy, _, _ := hs.isHealthy()
-	return isHealthy
-}
+// isHealthy returns the ProxyHealthzServer health state, lastUpdated and currentTime.
+func (hs *ProxyHealthzServer) isHealthy() (bool, time.Time, time.Time) {
+	lastUpdated, currentTime := time.Now(), time.Now()
```

Comment:
These need to be `hs.clock.Now()` in order to be properly comparable against the `proxierHealthManagers`.